### ISSUE TITLE
how to build pyodide's  pymupdf

### ISSUE THREAD

**Issue by burke-up** [pymupdf/PyMuPDF#2934]
### Description of the bug

i run the actions [Build wheels]
i set the wheels_linux_pyodide, but when i run i got the error:

"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/operations/prepare.py", line 525, in prepare_linked_requirement      
        return self._prepare_linked_requirement(req, parallel_builds)           
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^           
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/operations/prepare.py", line 640, in _prepare_linked_requirement     
        dist = _get_prepared_distribution(                                      
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^                                      
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/operations/prepare.py", line 71, in _get_prepared_distribution       
        abstract_dist.prepare_distribution_metadata(                            
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/distributions/sdist.py", line 54, in prepare_distribution_metadata   
        self._install_build_reqs(finder)                                        
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/distributions/sdist.py", line 124, in _install_build_reqs            
        build_reqs = self._get_build_requires_wheel()                           
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                           
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/distributions/sdist.py", line 101, in _get_build_requires_wheel      
        return backend.get_requires_for_build_wheel()                           
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                           
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_internal/utils/misc.py", line 751, in get_requires_for_build_wheel            
        return super().get_requires_for_build_wheel(config_settings=cs)         
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^         
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_vendor/pyproject_hooks/_impl.py", line 166, in get_requires_for_build_wheel   
        return self._call_hook('get_requires_for_build_wheel', {                
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_vendor/pyproject_hooks/_impl.py", line 321, in _call_hook                     
        raise BackendUnavailable(data.get('traceback', ''))                     
    pip._vendor.pyproject_hooks._impl.BackendUnavailable: Traceback (most recent
call last):                                                                     
      File                                                                      
"/home/runner/work/PyMuPDF/PyMuPDF/venv_pyodide/lib/python3.12/site-packages/pip
/_vendor/pyproject_hooks/_in_process/_in_process.py", line 77, in _build_backend
        obj = import_module(mod_path)                                           
              ^^^^^^^^^^^^^^^^^^^^^^^                                           
      File                                                                      
"/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/importlib/__init__.py",  
line 90, in import_module                                                       
        return _bootstrap._gcd_import(name, package, level)                     
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^             
      File "<frozen importlib._bootstrap>", line 1387, in _gcd_import           
      File "<frozen importlib._bootstrap>", line 1360, in _find_and_load        
      File "<frozen importlib._bootstrap>", line 1310, in                       
_find_and_load_unlocked                                                         
      File "<frozen importlib._bootstrap>", line [488](https://github.com/burke-up/PyMuPDF/actions/runs/7295762926/job/19882605967#step:7:489), in                        
_call_with_frames_removed                                                       
      File "<frozen importlib._bootstrap>", line 1387, in _gcd_import           
      File "<frozen importlib._bootstrap>", line 1360, in _find_and_load        
      File "<frozen importlib._bootstrap>", line 1331, in                       
_find_and_load_unlocked                                                         
      File "<frozen importlib._bootstrap>", line 935, in _load_unlocked         
      File "<frozen importlib._bootstrap_external>", line 994, in exec_module   
      File "<frozen importlib._bootstrap>", line 488, in                        
_call_with_frames_removed                                                       
      File                                                                      
"/tmp/pip-build-env-2xhg2bf5/overlay/lib/python3.12/site-packages/setuptools/__i
nit__.py", line 10, in <module>                                                 
        import distutils.core                                                   
    ModuleNotFoundError: No module named 'distutils'                            
                                                                                
                                                                                
Traceback (most recent call last):
  File "/home/runner/work/PyMuPDF/PyMuPDF/scripts/gh_release.py", line 645, in <module>
    main()
  File "/home/runner/work/PyMuPDF/PyMuPDF/scripts/gh_release.py", line 112, in main
    build(valgrind=valgrind)
  File "/home/runner/work/PyMuPDF/PyMuPDF/scripts/gh_release.py", line 198, in build
    build_pyodide_wheel(inputs_wheels_implementations)
  File "/home/runner/work/PyMuPDF/PyMuPDF/scripts/gh_release.py", line 385, in build_pyodide_wheel
    run(command, env_extra=env_extra)
  File "/home/runner/work/PyMuPDF/PyMuPDF/scripts/gh_release.py", line 604, in run
    subprocess.run(command, check=check, shell=1, env=env)
  File "/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/subprocess.py", line 571, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'true && echo "### cloning emsdk.git" && git clone https://github.com/emscripten-core/emsdk.git emsdk && echo "### creating venv venv_pyodide" && /opt/hostedtoolcache/Python/3.12.1/x64/bin/python -m venv venv_pyodide && . venv_pyodide/bin/activate && echo "### running pip install ..." && python -m pip install --upgrade pip wheel pyodide-build==0.23.4 && cd emsdk && PYODIDE_EMSCRIPTEN_VERSION=$(pyodide config get emscripten_version) && echo "### running ./emsdk install" && ./emsdk install ${PYODIDE_EMSCRIPTEN_VERSION} && echo "### running ./emsdk activate" && ./emsdk activate ${PYODIDE_EMSCRIPTEN_VERSION} && echo "### running ./emsdk_env.sh" && . ./emsdk_env.sh && ../wasm-opt-replace.py && cd .. && pyodide build --exports pyinit' returned non-zero exit status 2.
Error: Process completed with exit code 1.

### How to reproduce the bug

i run the actions [Build wheels]
set the wheels_linux_pyodide

### PyMuPDF version

1.23.8

### Operating system

Linux

### Python version

3.11

**Comment by julian-smith-artifex-com** [comment]
Interesting. I've just done a local pyodide build (not on github) and it worked fine for me.

I notice that you say that you're using python-3.11, but the output you include seems to be for python-3.12. And it looks like the core error is a failure of setuptools to import distutils:
```
File
"/tmp/pip-build-env-2xhg2bf5/overlay/lib/python3.12/site-packages/setuptools/init.py", line 10, in import distutils.core ModuleNotFoundError: No module named 'distutils'
```
Python-3.12 has removed support for distutils - see https://docs.python.org/3/whatsnew/3.12.html. Maybe there's a problem on your system where setuptools hasn't been updated properly to cope with this?

Could you try building with python-3.11?

**Comment by julian-smith-artifex-com** [comment]
I've just added a Github action to build a Pyodide wheel once per day. It works fine with python-3.11, but i get the same setuptools/distutils error as above with python-3.12.

See: .github/workflows/test_pyodide.yml

**Comment by burke-up** [comment]
> I've just added a Github action to build a Pyodide wheel once per day. It works fine with python-3.11, but i get the same setuptools/distutils error as above with python-3.12.
> 
> See: .github/workflows/test_pyodide.yml

thanks for your reply. it really succ. but when i use it in pyodide, i got the errror: AttributeError: 'float' object has no 
attribute 'm_internal'

**How to reproduce the bug**:
import pyodide.http
r = await pyodide.http.pyfetch('https://pymupdf.io/docs/mupdf_explored.pdf')
data = await r.bytes()
doc = fitz.Document(stream=data)
print(f'Is PDF: {doc.is_pdf}')
print(f'Number of pages: {doc.page_count}')

import json
page=doc[0]
page_json_str =doc[0].get_text("json")
page_json_data = json.loads(page_json_str)
span=page_json_data.get("blocks")[0].get("lines")[0].get("spans")[0]
page.add_redact_annot(span["bbox"],  text="")
page.apply_redactions()



**Detail errors**:

Traceback (most recent call last):
  File "<console>", line 1, in <module
>
  File "/lib/python3.11/site-packages/
fitz/__init__.py", line 8030, in add_r
edact_annot
    annot = self._add_redact_annot(qua
d, text=text, da_str=da_str,
            ^^^^^^^^^^^^^^^^^^^^^^^^^^
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/
fitz/__init__.py", line 7275, in _add_
redact_annot
    q = JM_quad_from_py(quad)
        ^^^^^^^^^^^^^^^^^^^^^
  File "/lib/python3.11/site-packages/
fitz/__init__.py", line 16446, in JM_q
uad_from_py
    if not obj.m_internal or not PySeq
uence_Check(obj) or PySequence_Size(ob
j) != 2:
           ^^^^^^^^^^^^^^
AttributeError: 'float' object has no 
attribute 'm_internal'

**Comment by julian-smith-artifex-com** [comment]
Thanks for reporting the details of the add_redact_annot bug. It's not specific to pyodide, and i've just pushed a fix.

**Comment by julian-smith-artifex-com** [comment]
I think this issue is fixed now. We'll probably want to be able to build for pyodide with python-3.12 in the future, but that will be managed separately.

**Comment by julian-smith-artifex-com** [comment]
The `add_redact_annot()` bug is fixed in release candidate 1.23.9rc1; see: #2960
